src/storage/index/hnsw_index.cpp

```diff
@@ -339,37 +338,272 @@ common::offset_t OnDiskHNSWIndex::searchNNInUpperLayer(transaction::Transaction*
+    NodeTableScanState& embeddingScanState, graph::NbrScanState& secondHopScanState,
+    min_node_priority_queue_t& candidates, max_node_priority_queue_t& results) const {
+    min_node_priority_queue_t candidatesForSecHop;
+
+    uint64_t numNbrsInFirstHop = 0;
+    uint64_t numVisitedNbrs = 0;
+    for (const auto& neighborChunk : nbrItr) {
+        neighborChunk.forEach([&](auto neighbor, auto) {
+            auto nbrOffset = neighbor.offset;
+            numNbrsInFirstHop++;
+            if (!visited.contains(nbrOffset)) {
+                const auto nbrVector =
+                    embeddings->getEmbedding(transaction, embeddingScanState, nbrOffset);
+                auto dist = HNSWIndexUtils::computeDistance(config.distFunc, queryVector, nbrVector,
+                    embeddings->getDimension());
+                candidatesForSecHop.push({nbrOffset, dist});
+                if (semiMask->isMasked(nbrOffset)) {
+                    if (results.size() < ef || dist < results.top().distance) {
+                        if (results.size() == ef) {
+                            results.pop();
+                        }
+                        numVisitedNbrs++;
+                        visited.add(nbrOffset);
+                        results.push({nbrOffset, dist});
+                        candidates.push({nbrOffset, dist});
+                    }
+                }
+            }
+        });
+    }
+
+    common::offset_vec_t filteredNeighbors;
+    while (!candidatesForSecHop.empty()) {
+        auto [candidate, candidateDist] = candidatesForSecHop.top();
+        candidatesForSecHop.pop();
+        if (numVisitedNbrs >= numNbrsInFirstHop) {
+            break;
+        }
+        if (visited.contains(candidate)) {
+            continue;
+        }
+        visited.add(candidate);
+        auto secondHopNbrItr =
+            lowerGraph->scanFwd(common::nodeID_t{candidate, nodeTableID}, secondHopScanState);
+        for (const auto& secondHopNbrChunk : secondHopNbrItr) {
+            secondHopNbrChunk.forEach([&](auto neighbor, auto) {
                 if (!visited.contains(neighbor.offset)) {
-                    const auto nbrVector =
-                        embeddings->getEmbedding(transaction, embeddingScanState, neighbor.offset);
-                    processNbrNodeInKNNSearch(queryVector, nbrVector, neighbor.offset, ef, visited,
-                        config.distFunc, embeddings.get(), candidates, result);
+                    if (semiMask->isMasked(neighbor.offset)) {
+                        visited.add(neighbor.offset);
+                        numVisitedNbrs++;
+                        filteredNeighbors.push_back(neighbor.offset);
+                    }
                 }
             });
         }
     }
-    return popTopK(result, k);
+    return filteredNeighbors;
+}
+
+common::offset_vec_t OnDiskHNSWIndex::blindTwoHopFilteredSearch(common::SemiMask* semiMask,
```

Comment:
I don't think this is the correct heuristic. What we want in blindTwoHop is this:
First iterate through each 1st degree neighbors and run dist only on the ones in mask (in the directed one you run the distance function on all 1st degree neighbors). 
Then, after all 1st degree neighbor traversal is done, iterate through each 2nd degree neighbors.

Also as above we should be doing this until M many distance calculations.

Reply:
I've changed the implementation to first iterate all 1st degree neighbours, then start with 2nd ones. I realized I was following the implementation of acron's blind search.